### ISSUE TITLE
질문드립니다.

### ISSUE THREAD

**Issue by KevinRyu** [seriousran/TMI-Deep-Learning#1]
도서 구입자입니다. 2가지 질문드리겠습니다. (tensorflow-gpu 2.2.0 환경에서 실습중입니다.)

질문1.

112페이지 fit_generator 함수를 쓰면 fit으로 사용하라고 나와서 fit으로 실행하면
Epoch 1/5
25/72 [=========>....................] - ETA: 1s - loss: 8.8672e-05 - accuracy: 1.0000 ETA: 1s - loss: 2.9520e-04 - accura - ETA: 1s - loss: 1.0843e-04 - accuracy: 1.WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches (in this case, 360 batches). You may need to use the repeat() function when building your dataset.
WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches (in this case, 72 batches). You may need to use the repeat() function when building your dataset.
25/72 [=========>....................] - 1s 34ms/step - loss: 8.8672e-05 - accuracy: 1.0000 - val_loss: 2.1379e-04 - val_accuracy: 1.0000

Data 부족으로 나옵니다. 플롯결과도 도서와 같이 정상적으로 출력 안됩니다. 어떻게 해결해야 하나요?

질문2. 
101페이지의 평가결과는 어떻게 해석해야 하는건가요? 출력 값만 나와있어서 의미를 잘 모르겠습니다.

**Comment by KevinRyu** [comment]
위의 에러에 대해

history = _model_1.fit(
    train_generator,
    epochs=5,
    steps_per_epoch=24,
    validation_data=validation_generator,
    validation_steps=6
)

이렇게 수정되야 하는것 아닌가요? 
이미지수는 786(train), 196(valid) 이런데요.

**Comment by KevinRyu** [comment]
train_datagen의 batch size를 수정해서 책과 같은 결과를 얻었습니다.

**Comment by seriousran** [comment]
안녕하세요!

우선, 이 책은 keras를 기반으로 집필되었지만, 현재는 tf에 포함되어 변화가 있는 상태입니다ㅠㅠ
말씀하신대로, fit_generator함수는 tensorflow 2.1부터 deprecated 되었습니다.

사실 가장 정확한 방법은

steps_per_epoch = len(train_images)//batch_size
validation_steps = len(valid_images)//batch_size

입니다. 이전의 keras에서는 부족한 만큼 더 repeat()해서 채우고 있지만, 현재는 default로는 채우지 않는 것으로 보여 보여주신 문제가 발생한 것으로 보입니다.

해당 코드에 적용해본다면,

steps_per_epoch =768//32
validation_steps = 196//32

가 될 것 같습니다!

(답변을 작성하는 도중에 먼저 문제를 해결하셨군요..!)

**Comment by KevinRyu** [comment]
감사합니다. 

3.4절의 Word 분류도.. 책의 예제 코드를 그대로 따라할 경우
Epoch 1/30
253/253 [==============================] - 1s 5ms/step - loss: 2.4082 - accuracy: 0.4459 - val_loss: 2.0068 - val_accuracy: 0.5161
Epoch 2/30
253/253 [==============================] - 1s 3ms/step - loss: 1.7510 - accuracy: 0.5608 - val_loss: 1.7784 - val_accuracy: 0.5840
Epoch 3/30
253/253 [==============================] - 1s 3ms/step - loss: 1.5634 - accuracy: 0.6113 - val_loss: 1.6690 - val_accuracy: 0.6129
Epoch 4/30

와 같이 학습되고, 8982개 x_train 데이터 수에서 0.1% validation set으로 분할하면 8083개 x_train..
이것을 batch size32로 나누면 253/253이 맞는데 도서에는 모두 batch_size로 나누기 전의 dataset 수로 적혀있습니다.
혼란을 줄수 있는 부분인 것 같습니다.

추가로, 제 환경에서 steps_per_epoch =786//32, validation_steps = 196//32 로 셋팅할경우 첫번째 학습결과부터 accuracy가 1.0이 나와서 책과같은 꺾인 그래프로 (정확도 그래프) 나오지 않고 flat한 그래프로 plot 됩니다. step을 72로 고정하고 batch_size를 줄여서 해야 책과 동일하게 됩니다. 

제 짧은 지식으로 이해가 100% 안되는 부분입니다.

**Comment by seriousran** [comment]
@KevinRyu 

해당 부분은 버전의 차이(keras에서 tensorflow로의 통합)으로 인한 차이입니다.
8083은 데이터의 수, 253은 iteration의 수로 생각하시면 될 것같습니다. 제가 그 숫자들을 직접 타이핑한 것이 아니라, 출력 결과를 그대로 복사 붙여넣기한 것이었습니다. 버전 문제로 혼란을 드리게되어 죄송합니다.

모델을 학습할 때에는 epoch 수와 batch_size를 설정합니다. 그리고 step은 부수적으로 설정되는 숫자로 생각하시면 좋을 것 같습니다. 일반적으로 batch_size가 높은 것이 안정적으로 잘 학습된다고 알려져 있습니다. batch_size가 8이라고 한다면, 8개의 데이터를 신경망을 통해 계산하고 정답과 비교하여 loss를 계산합니다. 그 loss를 바탕으로 가중치를 업데이트하여 모델을 학습합니다. batch_size가 16이라면 한번의 학습에 사용되는 데이터의 수는 16개죠. 한번에 여러 데이터를 갖고 loss를 계산해야 조금 더 다양한 경우에 대해서 한번에 입력이 되어 학습되기 때문에- 편향되지 않고 잘 학습될 가능성이 높습니다. (무조건적은 아니고, 예외 상황과 다른 요인들과도 관계가 있습니다.)

첫번째 학습결과(epoch)부터 accuracy가 1.0이 나왔다면, 학습이 매우잘된 결과입니다. batch_size를 줄여야 꺾인 그래프로 나온다는 의미는 학습이 한번에 잘된 것이 아니라, 여러번의 epoch를 통해서 학습이 수렴한 것으로 보입니다.

책과 동일하게 하였을 때 다른 결과가 나오는 것은 여러 요인이 존재할 수 있습니다. (랜덤 시드 설정, 코드에서 구체적으로 설정하지 않은 default값의 차이 등) 따라서, 분명하게 답변드리지 못하는 점 죄송합니다. accuracy가 높게 나오는 것으로 보았을 때, 학습은 잘 진행된 것으로 보입니다. 따라서 다른 부분에서 문제가 있지는 않은 것으로 보이므로 걱정하지 않고 넘어가셔도 좋을 것 같습니다. 출력결과가 다른 부분에 있어서 혼란을 드리어 죄송합니다. 딥러닝의 특성상 패키지 업데이트 속도가 너무 빠르고, 많은 변화가 이루어집니다. 또한, 랜덤시드를 고정하지 않는 경우가 있기 때문에- 같은 코드를 반복해서 실행하더라도 결과가 변동성이 있기도 합니다. 그런 점을 고려해주시면 감사드리겠습니다.

Batch_size의 차이로 인해서 생기는 변화는 매우 중요한 부분이기 때문에, 추가적으로 공부해서라도 이해하시고 넘어가시면 좋을 것 같습니다.

**Comment by KevinRyu** [comment]
상세한 설명 감사드립니다. 빠른 버젼업에, 버젼차이에 따른 문제는 저도 현업에서 고생하면서 충분히 이해하고 있습니다. ^^ 제가 버젼을 낮추어 테스트 하는 방법도 있으나 올드 버젼만 사용할수는 없으므로 힘들지만 찾아보고 해결하면서 하고 있습니다. 집필해주신 책은 튜토리얼이 아닌 현업에서 고민 할만한 부분들이 설명되어있어 정말 좋습니다. 다시한번 답변 감사합니다. close 하겠습니다.
